Refactor video plugin

7Ds7 opened the pull request:
After checking the PR #2110 i decided to take a look at Shaarli codebase to get myself familiar with it and check if i could help in anyway.

For the project to run properly on my machine i had to update a few packages (both composer and yarn) which might be a good thing as now the packages pass the audits (outdated libraries with CVEs)

Indeed the mentioned plugin although simple it is quite obfuscated with minified versions of other plugins within it, i usually do not use this plugin in my instance  but i think this refactored version has the same functionality as the current one.

- adds a button to the top bar to play videos
- creates a "modal" with a player
- lists next/prev in the videos 
- closes the modal

TLDR: This PR refactors the video plugin to a more modernized and non obfuscated code as well as updates packages that were outdated and/or failed audits.

thican commented:
Hello, thanks for your contribution, does it also work with an hardened CSP, as you saw in PR #2110?

thican commented on plugins/playvideos/youtube_playlist.js:
```diff
@@ -1,224 +1,185 @@
+    }
+  };
+
+}
```
Can it be ended with a new line please as most text files should?

thican commented on phpunit.xml:
```diff
@@ -1,8 +1,10 @@
-    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-    xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/4.5/phpunit.xsd"
-    colors="true">
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd" colors="true">
```
We could keep the attributes on different lines so it reduces the length line and the commit print.

7Ds7 replied:
On a different [PR](https://github.com/shaarli/Shaarli/pull/2140) now

thican reviewed (commented):
Also can you squash some commits, as the revert only fix the modification introduced in this PR.

nodiscc reviewed (changes requested):
> updates packages that were outdated and/or failed audits.

Please keep these unrelated changes for a separate PR, it makes reviewing easier (I think you could just discard/revert 99499b91672536fce6f0c262e289e3bd9308a6a8)

7Ds7 commented:
> > updates packages that were outdated and/or failed audits.
> 
> Please keep these unrelated changes for a separate PR, it makes reviewing easier (I think you could just discard/revert [99499b9](https://github.com/shaarli/Shaarli/commit/99499b91672536fce6f0c262e289e3bd9308a6a8))

Made a separate [PR](https://github.com/shaarli/Shaarli/pull/2140) for the packages

thican commented on plugins/playvideos/youtube_playlist.js:
```diff
@@ -1,224 +1,184 @@
+  }
+
+  getAllVideos = () => {
+    const reg = new RegExp("https?://(www.)?youtube.com/");
```
Does it also work with youtu.be links?

Also, the URL API might be better suited for this job.
https://developer.mozilla.org/en-US/docs/Web/API/URL_API

It could be possible a malicious link might have the whole `https://www.youtube.com/` in its path or in its query and then trigger some flaw by playing malicious video.

7Ds7 replied:
This as well as the [next comment](https://github.com/shaarli/Shaarli/pull/2138#discussion_r2049778072) are in reference to parts of the old plugin i left the functionality as it was  this was basically a port to get rid of jquery and bring the plugin to a more modern standard.

But you are right there is room for extended improvement just not sure if its place is in this PR

thican replied:
Okay I see, I thought it was a full new implementation as a lot of code were modified.

thican replied:
Just to ask, can you consider providing this extended improvement later, as you did already a large modification of this code?

thican commented on plugins/playvideos/youtube_playlist.js:
```diff
@@ -1,224 +1,184 @@
+      if (!$elm.getAttribute("href").match(reg)) {
+        return
+      }
+      const n = $elm.href.replace(/^.*v=/, "").replace(/\&.*$/, "");
```
Same here, I would have use the URL API to pick easily and safely the `v` value from the query part.

For youtu.be links, this is the path.

7Ds7 replied:
See reply [here](https://github.com/shaarli/Shaarli/pull/2138#discussion_r2049816459)

thican commented on plugins/playvideos/youtube_playlist.js:
```diff
@@ -1,224 +1,184 @@
+    this.getAllVideos();
+    this.createDom();
+
+    if (!document.head.querySelector('script[src="//www.youtube.com/iframe_api"]')) {
+      this.getScript('//www.youtube.com/iframe_api');
```
I would consider using explicit HTTPS links, as this is better nowadays as a security, and because Youtube answers `HTTP/1.1 301 Moved Permanently` to HTTP requests without TLS anyway.

7Ds7 replied:
`//` is https by default just like the [original plugin](https://github.com/shaarli/Shaarli/blob/master/plugins/playvideos/youtube_playlist.js#L89)

thican replied:
From my experience, which might be outdated, unspecified scheme in an URL is the same than the page, so HTTPS for HTTPS page, but HTTP for HTTP page.
I think we should not follow exactly what did the plugin when we notice an issue, hence the refactor/rework.

thican commented on plugins/playvideos/youtube_playlist.js:
May you please extra check with a linter? There is some errors like pasted below and some minor issues like trailing spaces, based on https://www.jslint.com/.

I don’t know if it is a recommended one however. I am not a JS dev so I can’t tell much more.
Also I can’t tell if this does the job, I am not able to currently test it.

Here some examples:
- `Use 'function (...)', not '(...) =>' when arrow functions become too complex.`,
- `Use double quotes, not single quotes.`,
- `Unexpected trailing space.`,
- `Naked block.`,
- `Expected ';' and instead […].`,
- `Unexpected 'this'.`,
- […]

7Ds7 replied:
There is no linter for javascript implemented in the Shaarli project as far as i am aware, the incorporated test suit passes. 

Not sure if it is a good idea to have a single js file linted from whatever opinion jslint.com has by default it might be a good idea though to implement a linter with accorded desired rules for javascript in the Shaarli test suite although it might be a big task given the that the project  codebase  never had a js linter before.

If needed i can gladly take a look at it and evaluate the effort, sadly i can only do it once this [PR](https://github.com/shaarli/Shaarli/pull/2140) or any other that updates the packages is merged because in the current state it does not run on my machine :(

thican replied:
"The current stage", you mean before this PR, or with? I don’t exactly recall, I guess it was working on mine when I was updating the CSP.

For the linter, I was mainly considering syntax issues. There is a main error where the online linter was not able to complete, I guess from a code issue like those missing semicolons.

7Ds7 replied:
Do you see any syntax errors on the console at runtime? if no then there are none

Upon a closer inspection it seems that  [eslint](https://github.com/shaarli/Shaarli/pull/2138#discussion_r2049778072) is already in place for but the linter does not run on the plugins folder my guess is because the linter was integrated after the plugins were added to the codebase 

`eslint:
	@yarnpkg run eslint -c .dev/.eslintrc.js assets/vintage/js/
	@yarnpkg run eslint -c .dev/.eslintrc.js assets/default/js/
	@yarnpkg run eslint -c .dev/.eslintrc.js assets/common/js/`

7Ds7 replied:
It is annoying to fix all these lint errors in which some of them are outdated and no longer part of eslint itself (nor the airbnb styles for that matter) because the packages are themselves old.

Although i just linted the file according to the in place eslint rules

```
 yarn run eslint -c .dev/.eslintrc.js plugins/playvideos/youtube_playlist.js
 172:5  warning  Unexpected console statement  no-console

✖ 1 problem (0 errors, 1 warning)
```

nodiscc replied:
There is no immediate need to lint these files, the old version of the plugin had not automatic linter, so this refactoring does not mandate one either. Sure it would be nice to have, but we can take care of this later/in a separate issue.

7Ds7 replied:
Yeah i was confused that @thican mentioned it...but anyway they are linted now according to the eslint in Shaarli

thican replied:
I had to mention it, every development nowadays should be done with a linter validation.

thican commented on plugins/playvideos/youtube_playlist.js:
```diff
@@ -1,224 +1,184 @@
-    run_playideos();
-});
+  getScript = url => new Promise((resolve, reject) => {
+    const script = document.createElement('script');
+    script.src = url;
+    script.async = true;
+    script.onerror = reject;
+    script.onload = script.onreadystatechange = function() {
```
Might be necessary to specify type, such as `script.type = "text/javascript";`.

thican replied:
Okay, I agree, omitted is fine then.

thican commented on plugins/playvideos/youtube_playlist.js:
```diff
@@ -1,224 +1,184 @@
+
+  destroyInstance = () => {
+    this.ytInstance.destroy()
+    document.querySelector('#player-background').remove();
```
Not sure, I would have use `getElementById` for an id, no?

7Ds7 replied:
A bit irrelevant, browsers already caught up with `querySelector` for quite some time now and it is not that we are doing huge DOM manipulations that `getElementById` would be perceived faster

thican replied:
I see, I used to barely use `querySelector` where I was not able to use `getElementById` or `getElementsByClass`.

thican commented on plugins/playvideos/README.md:
```diff
@@ -30,10 +30,6 @@ File: youtube_playlist.js
 Copyright: (c) 2010-2014, David Kelso <david@kelso.id.au>
```
Guess you could update your name here for contributing, no?

thican commented on plugins/playvideos/youtube_playlist.js:
```diff
@@ -1,224 +1,184 @@
+  createDom = () => {
+    let $bg = document.createElement('div');
+    $bg.id = 'player-background';
+    $bg.style.position = 'fixed';
+    $bg.style.top = 0;
+    $bg.style.left = 0;
+    $bg.style.display = 'flex';
+    $bg.style.justifyContent = 'center';
+    $bg.style.alignItems = 'center';
+    $bg.style.zIndex = 1001;
+    $bg.style.width = '100%';
+    $bg.style.height = '100%';
+    $bg.style.backgroundColor = 'rgba(0,0,0,.8)';
```
Not raising an issue, at lease for now on, I hope it will work with an hardened CSP.

7Ds7 replied:
Not sure if i follow

nodiscc replied:
@7Ds7 
@thican submitted changes that would allow for a hardened Content-Security-Policy setting in https://github.com/shaarli/Shaarli/pull/2110, I guess what he meant by this is that he hopes the new version of the playvideos plugin will also be compliant. We can see about that later, one thing at a time.

7Ds7 replied:
It is the same as long as one adds the policy to the headers all is good

nodiscc reviewed (commented):
Thank you, great work rewriting this plugin (basically from scratch?)
I have looked at the diff and see nothing wrong, however this needs to be tested, I will run some basic tests when time allows (other testers are also welcome)

Regarding feature parity with the old version of the plugin, not much is expected, click the plugin button, a video embed/overlay appears with working video controls, next/previous buttons, that's about it

nodiscc reviewed (approved):
Works fine, thank you!
There's an old issue about adding soundcloud (and/or other hosts) support to that plugin, if you feel up for it https://github.com/shaarli/Shaarli/issues/1454
![image](https://github.com/user-attachments/assets/f4ee4669-803f-47fc-87ad-4797943d2e45)
feelsgood.jpg

nodiscc reviewed (changes requested):
I'm sorry, one last thing. Can you please drop `345a61228282c2c4ae487cb150bffe33b362c995` (merge commit) so that this PR only contains the 2 relevant commits?

I'd like to keep the history as linear as possible (I have also considered switching to rebasing/squashing all PRs but that's for another day)

nodiscc commented on plugins/playvideos/README.md:
```diff
@@ -30,10 +30,6 @@ File: youtube_playlist.js
 Copyright: (c) 2010-2014, David Kelso <david@kelso.id.au>
 License: The ISC License (http://opensource.org/licenses/ISC)
```
```suggestion
```

Let's remove this and put this rewritten plugin under the global zlib license https://github.com/shaarli/Shaarli/blob/master/COPYING#L1 do you agree?

thican commented:
> I'd like to keep the history as linear as possible (I have also considered switching to rebasing/squashing all PRs but that's for another day)

I also prefer rebasing but only done by the main author, so stuff like signatures remain valid.

thican commented:
> I'm sorry, one last thing. Can you please drop `345a61228282c2c4ae487cb150bffe33b362c995` (merge commit) so that this PR only contains the 2 relevant commits?

And even the "2 relevant commits" should 100% be squashed.

7Ds7 commented:
@nodiscc @thican 

The history for this plugin was already lost more than once due to all the squashing, i am pretty sure that there are better flows than this, i am in favor of not loosing any history and i am in favor of merging the main branch often but that is on my own/team projects. Maybe there should be a consideration of adding some of these code guidelines upon [here](https://shaarli.readthedocs.io/en/master/dev/Development.html) ? or maybe it was me that did not found them and they do exist?

Same goes for linting, and linting rules, this already broke the plugin due to the enforce of 'class-methods-use-this' (my bad for not spotting it before)  in which i do not think it should be enforced at all due to the nature of the project i just spotted it and disabled this rule inline and removed the static methods from the class [here](https://github.com/shaarli/Shaarli/pull/2138/files#diff-845a75fe05cb516dae42e08e82fc2695ed9c7d3f0a7ace7f6e88d1acd3f22e51R180-R202)

Regarding the [soundcloud/vimeo](https://github.com/shaarli/Shaarli/pull/2138#pullrequestreview-2795408220) situation, i added for now at least, the shortened youtu.be urls and sort of prepared code-wise so that we can add more.

There seems to be [soundcloud widget](https://developers.soundcloud.com/docs/api/html5-widget) and a [vimeo sdk](https://developer.vimeo.com/player/sdk/basics) that we can integrate and they both seem to have a event for "media ended" that we can use to keep playing the media after this one is ended.
I can probably take a look at it when i have the time but i would do it on another pull request and probably after some discussion about it, some questions that come from the top of my mind are:

- will the plugin be replaced for example media_playlist?
- will the users that already use youtube playlist feel awkward that will now play other media?
- should this services be opt-in ex: the user chooses which they want to be included ? (might be a bit of a pickle to do that)

nodiscc commented:
Discussions about merge style should go in a separate issue, let's not overcomplicate this one. Thanks for dropping the merge commit.

Regarding possible vimeo/soundcloud integration, I wll reply in https://github.com/shaarli/Shaarli/issues/1454

Thanks again
